Arrowgene.Ddon.GameServer/InstanceAssetManager.cs

```diff
@@ -1,53 +1,90 @@
-using System.Collections.Generic;
-using Arrowgene.Ddon.Shared;
 using Arrowgene.Ddon.Shared.Entity.Structure;
 using Arrowgene.Ddon.Shared.Model;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace Arrowgene.Ddon.GameServer.GatheringItems
 {
-    public abstract class InstanceAssetManager<T1, T2, T3>
+    public abstract class InstanceAssetManager<TItem, TAssetItem>
     {
         public InstanceAssetManager()
         {
-            this._instancedAssetsDictionary = new Dictionary<(StageId, T1), List<T3>>();
+            this._instancedAssetsDictionary = new Dictionary<StageId, Dictionary<int, TAssetItem>>();
+        }
+
+        private readonly Dictionary<StageId, Dictionary<int, TAssetItem>> _instancedAssetsDictionary;
+
+        public bool HasAssetsInstanced(CDataStageLayoutId stageLayoutId, int subId)
+        {
+            var stageId = StageId.FromStageLayoutId(stageLayoutId);
+            return HasAssetsInstanced(stageId, subId);
         }
 
-        private readonly Dictionary<(StageId, T1), List<T3>> _instancedAssetsDictionary;
+        public bool HasAssetsInstanced(StageId stageId, int subId)
+        {
+            return _instancedAssetsDictionary.ContainsKey(stageId) && _instancedAssetsDictionary[stageId].ContainsKey(subId);
+        }
 
-        public bool HasAssetsInstanced(CDataStageLayoutId stageLayoutId, T1 subGroupId)
+        public List<TAssetItem> GetAssets(CDataStageLayoutId stageLayoutId)
         {
-            return _instancedAssetsDictionary.ContainsKey((StageId.FromStageLayoutId(stageLayoutId), subGroupId));
+            return GetAssets(StageId.FromStageLayoutId(stageLayoutId));
         }
 
-        public bool HasAssetsInstanced(StageId stageId, T1 subGroupId)
+        public List<TAssetItem> GetAssets(StageId stageId)
         {
-            return _instancedAssetsDictionary.ContainsKey((stageId, subGroupId));
+            if (!_instancedAssetsDictionary.ContainsKey(stageId))
+            {
+                IEnumerable<TItem> items = FetchAssetsFromRepository(stageId);
+                List<TAssetItem> instancedAssets = InstanceAssets(items);
+                if (!_instancedAssetsDictionary.ContainsKey(stageId))
+                {
+                    _instancedAssetsDictionary[stageId] = new Dictionary<int, TAssetItem>();
+                }
+                for (int i = 0; i < instancedAssets.Count; i++)
+                {
+                    _instancedAssetsDictionary[stageId].Add(i, instancedAssets[i]);
+                }
+                return instancedAssets;
+            }
+            else
+            {
+                return _instancedAssetsDictionary[stageId].Values.ToList();
+            }
         }
 
-        public List<T3> GetAssets(CDataStageLayoutId stageLayoutId, T1 subGroupId)
+        public TAssetItem GetAssets(CDataStageLayoutId stageLayoutId, int subId)
```

Comment:
Why did you rename subgroup to sub?

Reply:
I see your point, yeah, we should write down this reasoning somewhere, could you add an explanation to the comments?